framework/src/userobject/RenormalizeVector.C

```diff
@@ -0,0 +1,123 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "RenormalizeVector.h"
+#include "MooseError.h"
+#include "NonlinearSystemBase.h"
+
+#include "libmesh/numeric_vector.h"
+#include "libmesh/int_range.h"
+
+registerMooseObject("MooseApp", RenormalizeVector);
+
+InputParameters
+RenormalizeVector::validParams()
+{
+  InputParameters params = GeneralUserObject::validParams();
+  params.addClassDescription(
+      "Pointwise renormalize the solution of a set of variables comprising a vector");
+  params.addCoupledVar("v", "Variables comprising the vector");
+  params.addParam<Real>("norm", 1.0, "Desired norm for the coupled variable vector");
+  return params;
+}
+
+RenormalizeVector::RenormalizeVector(const InputParameters & parameters)
+  : GeneralUserObject(parameters),
+    _mesh(_fe_problem.mesh()),
+    _var_names(getParam<std::vector<VariableName>>("v")),
+    _target_norm(getParam<Real>("norm")),
+    _nl_sys(_fe_problem.getNonlinearSystemBase()),
+    _sys(_nl_sys.system())
+{
+  for (const auto & var_name : _var_names)
+  {
+    auto & var = _fe_problem.getVariable(0, var_name);
+    if (_sys.number() != var.sys().system().number())
+      paramError("v", "Variables must be all in the non-linear system.");
+
+    if (var.isArray())
+    {
+      const auto & array_var = _fe_problem.getArrayVariable(0, var_name);
+      for (unsigned int p = 0; p < var.count(); ++p)
+        _var_numbers.push_back(_sys.variable_number(array_var.componentName(p)));
+    }
+    else
+      _var_numbers.push_back(_sys.variable_number(var_name));
+  }
+}
+
+void
+RenormalizeVector::initialize()
+{
+  // do one solution.close to get updated
+  _sys.solution->close();
+}
+
+void
+RenormalizeVector::execute()
+{
+  auto & dof_map = _sys.get_dof_map();
+  const auto local_dof_begin = dof_map.first_dof();
+  const auto local_dof_end = dof_map.end_dof();
+
+  std::vector<std::vector<dof_id_type>> dof_indices(_var_numbers.size());
+  std::vector<Real> cache(_var_numbers.size());
+
+  for (const auto & elem : *_mesh.getActiveLocalElementRange())
+  {
+    // prepare variable dofs
+    for (const auto i : index_range(_var_numbers))
+    {
+      dof_map.dof_indices(elem, dof_indices[i], _var_numbers[i]);
+
+      // check that all vars have the same number of dofs
+      mooseAssert(dof_indices[i].size() == dof_indices[0].size(),
```

Comment:
this should be a regular error UNLESS you check in the construction:
- all variables have the same block restriction
- all variables have the same number of DOFs per element (so same variable FEType will do)

if we dont pretty sure user will get a seg fault a few lines below with the non-matching variables